### ISSUE TITLE
Color Configuration issues

### ISSUE THREAD

**Issue by jens-maus** [jens-maus/yam#162]
**Originally by _opionline@yahoo.de_ on 2010-04-21 15:25:16 +0200**

___
The config->Read page have 2 problems:

1. If "Show text colors" is disabled in the config the poppen buttons should be disabled as well if the page is opened.

2. At least on OS3.x the colors ar not displayed in the poppen buttons if the page is opened. See also:
http://faq.yam.ch/content/1/5/en/why-does-the-colored-text-gadget-in-the-read-configuration-looks-wrong-the-first-time-i-open-this-sheet.html

The attached patch YAM_COg.diff should fix both issues.

The question is should we enclose the "set(data->GUI..." calls into a
```
#if defined(__amigaos4__) || defined(__MORPHOS__) || defined(__AROS__)
#endif
```
pair so it's only applied on OS3.x?

To save #defined statements it could be also applied to all systems and should
not cause negative effects and the execution time should be very fast too.

any suggestions?

**Comment by jens-maus** [comment]
**Originally by _opionline@yahoo.de_ on 2010-04-21 15:25:29 +0200**

___
**Attachment added:** [`YAM_COg.diff`](https://yam.ch/attachments/162/YAM_COg.diff) (2.3 KiB)

**Comment by tboeckel** [comment]
**Originally on 2010-04-21 16:16:02 +0200**

___
1. The problem is MUI itself. At creation time the checkmark object has MUIA_Selected==FALSE. When YAM displays the config page it sets MUIA_Selected=FALSE a second time. This set() call is optimized away by MUI and hence no notification is issued.
Since all the CO_Page#?() functions only create the individual pages and set up some notifications I suggest to move the MUIM_MultiSet method call to CO_SetConfig() in YAM_COs.c. Notifications only work for real changes and for user input. Without a change no notification will be triggered.
A different workaround would look like this:
      setcheckmark(gui->CH_TEXTCOLORS_READ, TRUE);
      setcheckmark(gui->CH_TEXTCOLORS_READ, CE->UseTextColorsRead);
Thus the notification always gets triggered, because either the first or the second setcheckmark() performs a real change.

2. This is an issue of MUI 3.8. MUI 3.9 seems to have this one fixed. MUI 3.8 seems to perform a similar optimization as above since setting the same pens a second time doesn't change anything. Your proposed change can also be put into CO_SetConfig() instead of CO_PageRead() to let the modification look like this:
      set(gui->CA_COLSIG,   MUIA_Pendisplay_Spec, &C->ColorSignature);
      set(gui->CA_COLSIG,   MUIA_Pendisplay_Spec, &CE->ColorSignature);
The first operation is only necessary for MUI 3.8, so this could be enclosed with an #ifdef, but on the other hand it should not hurt if all systems do this.

**Comment by jens-maus** [comment]
**Originally by _opionline@yahoo.de_ on 2010-04-21 17:00:22 +0200**

___
1. Some other CO_Page#?() functions also call set(...) for setting
some attributes. But i can try to move it into CO_SetConfig() if it's
the better place for such things.

2. Your suggestion may work but need 2 set() calls for every poppen.
Is it really needed?

let's see what Jens say about that.

**Comment by jens-maus** [comment]
**Originally by _opionline@yahoo.de_ on 2010-04-21 18:01:05 +0200**

___
Seems i should into the source before i reply.

1.
setcheckmark(gui->CH_TEXTCOLORS_READ, TRUE); 
setcheckmark(gui->CH_TEXTCOLORS_READ, CE->UseTextColorsRead);
does work fine here. :-)

2.
Your suggestion using the C->Color... does not work here
because both have the same value and does not change anything
so no notification is triggered. :-(

**Comment by jens-maus** [comment]
**Originally on 2010-04-21 18:10:54 +0200**

___
Well, I more or less have the same opinion like Thore.

 1. Move the MUIM_MultiSet method call to CO_SetConfig() in YAM_COs.c and do not use the workaround by setting the attribute to TRUE and immediately back to UseTextColorsRead. This looks like a bad workaround and should be avoided.

 2. Have you tried to do something like? :
```
set(gui->CA_COLSIG, MUIA_Pendisplay_Spec, NULL);
set(gui->CA_COLSIG, MUIA_Pendisplay_Spec, &CE->ColorSignature);
```
However, I would definitly enclose the NULL set into #ifdef and make a large comment that this has been added only because MUI 3.8 is buggy.

**Comment by jens-maus** [comment]
**Originally by _opionline@yahoo.de_ on 2010-04-21 19:26:34 +0200**

___
1.
Ok changed again and it works too.
I attach the resulting patch as YAM_COs.diff
but it's not yet complete because of:

2.
Yes of course that was the second try as i saw it was not working.
Seems MUI reject a null pointer here. :-(


Still searching for a (second working) solution...


BTW: Thanks for fixing the wrong formatting here.
i think it was because of some characters in one line!?

**Comment by jens-maus** [comment]
**Originally by _opionline@yahoo.de_ on 2010-04-21 19:27:22 +0200**

___
**Attachment added:** [`YAM_COs.diff`](https://yam.ch/attachments/162/YAM_COs.diff) (1.4 KiB)

**Comment by jens-maus** [comment]
**Originally by _opionline@yahoo.de_ on 2010-04-24 03:48:29 +0200**

___
I found a second (better?) solution for the MUI 3.8 Bug now.

The full patch for this bug would look like the attached file
Bug162.diff.

Maybe we could always activate the page before we call
CO_SetConfig() but i think it would slow down the display time
a lot on some pages so the patch is only applied on page cp_Read
which is currently the onl yone with Poppen Objects.

Is it ok if i commit the patch?

**Comment by jens-maus** [comment]
**Originally by _opionline@yahoo.de_ on 2010-04-24 03:48:54 +0200**

___
**Attachment added:** [`Bug162.diff`](https://yam.ch/attachments/162/Bug162.diff) (2.8 KiB)

**Comment by tboeckel** [comment]
**Originally on 2010-04-24 20:59:39 +0200**

___
Well, your patch is short, hopefully fixes this issue, includes a big fat warning and applies for one single system only. I'd say you can check it in as it is.

**Comment by jens-maus** [comment]
**Originally by _opionline@yahoo.de_ on 2010-04-24 22:05:50 +0200**

___
Ok i'll commit the patch ASAP.

Maybe you have a 3.x System to test if the issue is fixed for you too!?
At least here on a A2000 with OS3.9 BB2 it works.

The above mentioned website could get a update then too...

**Comment by jens-maus** [comment]
**Originally by _opionline@yahoo.de_ on 2010-04-24 22:33:05 +0200**

___
In (a9620b9):
   * YAM_CO.c: Add a workaround for a MUI 3.8 bug to show the colors for the
    Poppen objects on OS3.x.
  - YAM_COs.c: Disable all Poppen objects according to the UseTextColorsRead
    setting and do some cosmetic changes. This closes #162.

**Comment by tboeckel** [comment]
**Originally on 2010-04-24 23:16:54 +0200**

___
Yes, I have a WinUAE setup at work. I will verify it on Monday.

**Comment by jens-maus** [comment]
**Originally on 2010-08-26 11:38:40 +0200**

___
reopened due to upcoming merge on 2.6p1 branch.

**Comment by tboeckel** [comment]
**Originally on 2010-08-26 14:27:18 +0200**

___
In (9414594):
 * YAM_CO.c: Add a workaround for a MUI 3.8 bug to show the colors for the
  Poppen objects on OS3.x.
* YAM_COs.c: Disable all Poppen objects according to the UseTextColorsRead
  setting and do some cosmetic changes. This closes #162.
